File: week6/Week6Day4/app/src/main/java/com/example/luisenriquez/week6day4/AppController.java
package com.example.luisenriquez.week6day4;

import android.app.Application;

import com.example.luisenriquez.week6day4.di.DaggerMainComponent;
import com.example.luisenriquez.week6day4.di.MainComponent;
import com.example.luisenriquez.week6day4.di.MainModule;

public class AppController extends Application
{
    private MainComponent MyMainComponent;

    public MainComponent getAppComponent()
    {
        if (MyMainComponent == null)
        {
            MyMainComponent = DaggerMainComponent.builder()
                                                .mainModule(new MainModule(this))
                                                .build();
        }
        return MyMainComponent;
    }
}

File: Week7/Week7Day4Test/app/src/main/java/com/example/luisenriquez/week7day4test/AppController.kt
package com.example.luisenriquez.week7day4test

import android.app.Application
import com.example.luisenriquez.week7day4test.di.application.ApplicationComponent
import com.example.luisenriquez.week7day4test.di.application.ApplicationModule
import com.example.luisenriquez.week7day4test.di.application.DaggerApplicationComponent

class AppController : Application()
{
    private lateinit var applicationComponent: ApplicationComponent

    fun getComponent() : ApplicationComponent
    {
        return if (this::applicationComponent.isInitialized)
        {
            applicationComponent
        }
        else
        {
            createApplicationComponent()
        }
    }

    private fun createApplicationComponent(): ApplicationComponent
    {
        applicationComponent = DaggerApplicationComponent.builder()
                .applicationModule(ApplicationModule(this))
                .build()

        return applicationComponent
    }
}